### PR TITLE
sx127x: add several NETOPT for GNRC LoRaWAN

### PR DESCRIPTION
**Pull request by jia200x**
### Contribution description
This PR add the following NETOPT to the SX127x driver:
- NETOPT_RANDOM: Get a random number from the LoRa device
- NETOPT_SYNCWORD: Get/set the LoRa PHY Syncword.
- NETOPT_RX_SYMBOL_TIMEOUT: Set the number of symbols the LoRa transceiver should wait before triggering an RX_TIMEOUT event, when set into RX_SINGLE mode.

These commits were cherry-picked from #11022 

### Testing procedure
Compile and flash the `drivers_sx127x` test:
```
BOARD=b-l072z-lrwan1 make -C tests/driver_sx127x all flash term
```

Then:
- Verify the `random` command give random numbers
- Use the `syncword` command to get and set the syncword.
    - Set a syncword (in decimal) with "syncword set <dec>"
    - Get the syncword (in hex) with "syncword get". Check both numbers are equivalent
    - Also, check the same hex number in 0x39 register with `register get 0x39`
- Use the `rx_timeout` command to set the RX symbol timeout
    - `rx_timeout set <dec>`
    - Read the content of the registers 0x1E and 0x1F (rx_timeout is 10 bits long and scattered in these 2 registers). The 2 MSB of the rx_timeout are the 2 LSB of the 0x1E register and the 8 LSB are in the 0x1F reg. Check the hex matches the introduced rx_timeout value.
E.g
```1561389398.684083;st-lrwan1-11;> rx_timeout set 50
1561389398.686839;st-lrwan1-11;New rx_timeout set
st-lrwan1-11;register get 0x1E
1561389453.166605;st-lrwan1-11;> register get 0x1E
1561389453.168076;st-lrwan1-11;[regs] 0x1E = 0xC4
st-lrwan1-11;register get 0x1F
1561389457.242929;st-lrwan1-11;> register get 0x1F
1561389457.245584;st-lrwan1-11;[regs] 0x1F = 0x32
```
note ((0xC4 & 0x3) << 8) | 0x32 == 0x32 => 50 decimal

### CODE REVIEW

**Comment by jia200x** [comment]
> Why not adding a commit with these changes in the test application ?

Because I thought it's too hacky for its own commit. I think it would be possible to write unit tests for these get/set options (and check values with sx127x_reg_read). I can do it at some point

**Comment by aabadie** [comment]
> Because I thought it's too hacky for its own commit

I don't get your point.
Replacing `sx127x_random` by the corresponding netdev get function is straight forward. For the 2 other commands, you provided it in the patch, so should be ok to integrate it.
It'll also give a way to build (and manually test) the new netopts.

**Comment by jia200x** [comment]
@aabadie done!

**Comment by jia200x** [comment]
this is a dependency of #11022

File: sys/include/net/netopt.h [hunk 1]
```diff
@@ -645,6 +645,23 @@ typedef enum {
      */
     NETOPT_LORAWAN_MIN_RX_SYMBOL,
 
+    /**
+     * @brief   (uint8_t*) phy layer syncword
+     */
+    NETOPT_SYNCWORD,
```
Comment (smlng) [inline sys/include/net/netopt.h]: is this LoRa specific? Then it should be named `NETOPT_LORA_SYNCWORD` or `NETOPT_LORAWAN_SYNCWORD` or even `NETOPT_LORAWAN_PHY_SYNCWORD`?

File: sys/include/net/netopt.h [hunk 2]
```diff
@@ -645,6 +645,23 @@ typedef enum {
...
+     * @brief  (uint32_t) Get a random value from the device
+     *
+     * Nothing happens when set
+     */
+    NETOPT_RANDOM,
+
+    /**
+     * @brief (uint8_t) Get or set the number of PHY symbols before assuming there's no data
+     */
+    NETOPT_RX_SYMBOL_TIMEOUT,
```
Comment (smlng) [inline sys/include/net/netopt.h]: same as above, may need prefix `LORAWAN`?

File: sys/include/net/netopt.h (hunk 1 above)
Comment (aabadie) [reply sys/include/net/netopt.h]: This is a general concept for synchronizing data transmission (see [wikipedia](https://en.wikipedia.org/wiki/Syncword)), so no this is not LoRa specific.

Comment (jia200x) [reply sys/include/net/netopt.h]: yup. In most 15.4 radios it's possible to change the SFD (Syncword) to a different value

File: sys/include/net/netopt.h (hunk 2 above)
Comment (aabadie) [reply sys/include/net/netopt.h]: same as above, this is not LoRaWAN specific, even if only if LoRa requires it for the moment in RIOT.

File: tests/driver_sx127x/main.c [hunk 3]
```diff
@@ -269,6 +271,37 @@ int listen_cmd(int argc, char **argv)
     return 0;
 }
 
+int syncword_cmd(int argc, char **argv)
+{
+    if(argc < 2) {
```
Comment (aabadie) [inline tests/driver_sx127x/main.c]: This file needs a round of uncrustify. There are spaces like this missing at some place. And there are extra spaces after cast types.

Comment (jia200x) [reply tests/driver_sx127x/main.c]: ok, I will

Comment (jia200x) [reply tests/driver_sx127x/main.c]: done!

**Comment by jia200x** [comment]
forgot to add netopts to `crosslayer/netopt.c`. Amended and squashed directly

**Comment by aabadie** [comment]
24527fa536ada45453edaf0c9ec7c0da948dffc8 should be squashed with 2775c8b44c76a572de708d1d0a9c9046810b6982 (no need for a separate commit).

> forgot to add netopts to crosslayer/netopt.c. Amended and squashed directly

Ok !

**Comment by aabadie** [comment]
There is a problem with 24527fa: it modifies unrelated stuff...

**Comment by aabadie** [comment]
24527fa536ada45453edaf0c9ec7c0da948dffc8 commit message is valid: it references `tests/at86rf2xx` whereas it touches `tests/driver_sx127x`

File: tests/driver_sx127x/main.c [hunk 4]
```diff
@@ -122,9 +122,11 @@ int random_cmd(int argc, char **argv)
     (void)argc;
     (void)argv;
 
-    netdev_t *netdev = (netdev_t *)&sx127x;
+    netdev_t *netdev = (netdev_t*) &sx127x;
```
Comment (aabadie) [inline tests/driver_sx127x/main.c]: Uncrustify must be run again on this file.

File: tests/driver_sx127x/main.c [hunk 5]
```diff
@@ -270,6 +272,39 @@ int listen_cmd(int argc, char **argv)
...
+
+    if (strstr(argv[1], "set") != NULL) {
+        if (argc < 3) {
+            puts("usage: syncword set <syncword>");
+            return -1;
+        }
+        syncword = atoi(argv[2]);
+        netdev->driver->set(netdev, NETOPT_SYNCWORD, &syncword,
+                            sizeof(syncword));
+        printf("New syncword set\n");
```
Comment (aabadie) [inline tests/driver_sx127x/main.c]: can be replaced by puts, or eventually print the value that has been set.

File: tests/driver_sx127x/main.c [hunk 6]
```diff
@@ -304,9 +339,38 @@ int channel_cmd(int argc, char **argv)
...
+    uint16_t rx_timeout;
+    if (strstr(argv[1], "set") != NULL) {
+        if (argc < 3) {
+            puts("usage: rx_timeout set <rx_timeout>");
+            return -1;
+        }
+        rx_timeout = atoi(argv[2]);
+        netdev->driver->set(netdev, NETOPT_RX_SYMBOL_TIMEOUT, &rx_timeout,
+                            sizeof(rx_timeout));
+        printf("New rx_timeout set\n");
```
Comment (aabadie) [inline tests/driver_sx127x/main.c]: can be replaced by puts

File: tests/driver_sx127x/main.c [hunk 7]
```diff
@@ -270,6 +272,39 @@ int listen_cmd(int argc, char **argv)
...
+        printf("Syncword: %02x\n", (int)syncword);
+        return 0;
+    }
+
+    if (strstr(argv[1], "set") != NULL) {
+        if (argc < 3) {
+            puts("usage: syncword set <syncword>");
+            return -1;
+        }
+        syncword = atoi(argv[2]);
```
Comment (aabadie) [inline tests/driver_sx127x/main.c]: It's very counter intuitive regarding the `get` subcommand: here the input value is interpreted as an int but the `get` subcommand returns it as an hex.
So it's either everything is entered/printed as int or everything is entered/printed as hex.

**Comment by jia200x** [comment]
done :)

**Pull request merged**